WireDomain/Sources/WireDomain/Notifications/NotificationServiceExtension.swift

```
@@ -0,0 +1,209 @@
+
+                let notificationPayload = try NotificationPayload(
+                    userInfo: request.content.userInfo
+                )
+
+                let userID = notificationPayload.userID
+
+                let rootComponent = try setupRootComponent(
+                    userID: userID,
+                    notificationHandler: contentHandler
+                )
+
+                let verifyUserSession = rootComponent.verifyUserSession
+                let startSyncingEvents: () async throws -> Void = {
+                    try await verifyUserSession.startSyncingEvents(
+                        eventID: notificationPayload.eventID
+                    )
+                }
+
+                try await verifyUserSession.verify(
+                    userID: userID,
+                    then: startSyncingEvents
+                )
+
+            } catch {
+                logError(error)
+                finishWithEmptyNotification()
+            }
+        }
+    }
+
+    public func serviceExtensionTimeWillExpire() {
+        logger.warn("new notification service will expire", attributes: .newNSE)
+        finishWithEmptyNotification()
+    }
+
+    // With the "filtering" entitlement, we can tell iOS to not display a user notification by passing empty content to
+    // the content handler. See https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_developer_usernotifications_filtering
+    private func finishWithEmptyNotification() {
+        logger.info("finishing without showing notification", attributes: .newNSE)
+        contentHandler?(.emptyNotification)
+        terminate()
+    }
+
+    private func setupRootComponent(
+        userID: UUID,
+        notificationHandler: @escaping (UNNotificationContent) -> Void
+    ) throws -> RootComponent {
+        let infoDictionary = Bundle.main.infoDictionary
+        guard let appGroupID = infoDictionary?["WireGroupId"] as? String else {
+            fatalError()
```

issue: it would be better to throw an error than crash.

---

yes 👍  https://github.com/wireapp/wire-ios/pull/2397/commits/fd2bef5257ac080234df890bf248f2e0f5e8a7d0